rust/src/qfunction.rs

```diff
@@ -0,0 +1,532 @@
+    }
+}
+/// View a QFunction
+///
+/// ```
+/// # use libceed::prelude::*;
+/// # let ceed = libceed::Ceed::default_init();
+/// let mut user_f = |
+///   q: usize,
+///   inputs: &[&[f64]],
+///   outputs: &mut [&mut [f64]],
+/// | -> i32
+/// {
+///   // Inputs
+///   let u = &inputs[0];
+///   let weights = &inputs[1];
+///   // Outputs
+///   let v = &mut outputs[0];
+///
+///   // Loop over quadrature points
+///   v.iter_mut()
+///     .zip(u.iter().zip(weights.iter()))
+///     .for_each(|(v, (u, w))| *v = u * w);
+///
+///   // Return clean error code
+///   0
+/// };
+///
+/// let mut qf = ceed.q_function_interior(1, Box::new(user_f));
+///
+/// qf.add_input("u", 1, EvalMode::Interp);
+/// qf.add_input("weights", 1, EvalMode::Weight);
+///
+/// qf.add_output("v", 1, EvalMode::Interp);
+///
+/// println!("{}", qf);
+/// ```
+impl fmt::Display for QFunction {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.qf_core.fmt(f)
+    }
+}
+
+/// View a QFunction by Name
+///
+/// ```
+/// # use libceed::prelude::*;
+/// # let ceed = libceed::Ceed::default_init();
+/// let qf = ceed.q_function_interior_by_name("Mass1DBuild");
+/// println!("{}", qf);
+/// ```
+impl fmt::Display for QFunctionByName {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.qf_core.fmt(f)
+    }
+}
+
+// -----------------------------------------------------------------------------
+// Core functionality
+// -----------------------------------------------------------------------------
+impl QFunctionCore {
+    // Common implementation
+    pub fn apply(&self, Q: i32, u: &[Vector], v: &[Vector]) {
+        let mut u_c = [std::ptr::null_mut(); crate::MAX_QFUNCTION_FIELDS];
+        for i in 0..std::cmp::min(crate::MAX_QFUNCTION_FIELDS, u.len()) {
+            u_c[i] = u[i].ptr;
+        }
+        let mut v_c = [std::ptr::null_mut(); crate::MAX_QFUNCTION_FIELDS];
+        for i in 0..std::cmp::min(crate::MAX_QFUNCTION_FIELDS, v.len()) {
+            v_c[i] = v[i].ptr;
+        }
+        unsafe {
+            bind_ceed::CeedQFunctionApply(self.ptr, Q, u_c.as_mut_ptr(), v_c.as_mut_ptr());
+        }
+    }
+}
+
+// -----------------------------------------------------------------------------
+// User QFunction Closure
+// -----------------------------------------------------------------------------
+pub type QFunctionUserClosure = dyn FnMut(usize, &[&[f64]], &mut [&mut [f64]]) -> i32;
+
+unsafe extern "C" fn trampoline(
+    ctx: *mut ::std::os::raw::c_void,
+    q: bind_ceed::CeedInt,
+    inputs: *const *const bind_ceed::CeedScalar,
+    outputs: *const *mut bind_ceed::CeedScalar,
+) -> ::std::os::raw::c_int {
+    let context = &mut *(ctx as *mut QFunction);
+    let trampoline_data = &context.trampoline_data;
+
+    // Inputs
+    let inputs_slice: &[*const bind_ceed::CeedScalar] =
+        std::slice::from_raw_parts(inputs, crate::MAX_QFUNCTION_FIELDS);
+    let mut inputs_vec: Vec<&[f64]> = Vec::new();
```

Comment:
Could we have arrays of slices here to prevent transforming `Vec` into arrays?

Reply:
I picked vec with a runtime determined length so the user couldn't reach into the inputs/outputs array into an uninitalized entry, but I don't feel strongly about it.

Reply:
I think what @jedbrown proposed with `ArrayVec` solves this issue.